### PR TITLE
fix(browser): wait for CDP readiness after start to avoid PortInUseError races (#21149)

### LINKED ISSUE #21149: [Bug]: browser start/open blocks snapshot/screenshot with PortInUseError on Linux (port 18800 conflict)
### Summary

Calling `browser(action="start")` or `browser(action="open")` before `browser(action="snapshot")` or `browser(action="screenshot")` causes a PortInUseError that makes all functional browser tool calls fail until gateway restart.

### Steps to reproduce

1. Call browser(action="start", profile="openclaw") or browser(action="open", targetUrl="...", profile="openclaw")
2. Call browser(action="snapshot", ...) or browser(action="screenshot", ...)
3. Error: "PortInUseError: Port 18800 is already in use"
4. All subsequent snapshot/screenshot/act/navigate calls fail with the same error
5. Restarting the gateway restores normal function

Alternatively: call snapshot or screenshot after any previous browser session left a stale Chromium process (e.g. after a gateway restart without pkill).

### Expected behavior

snapshot/screenshot/act should work regardless of whether start or open was called first. The browser control service should either reuse the existing Chromium instance or handle the port conflict gracefully.

### Actual behavior

Calling start or open launches Chromium directly with --remote-debugging-port=18800. When snapshot or screenshot is subsequently called, the browser control service attempts to start its own Chromium on the same port and fails with:

"Can't reach the OpenClaw browser control service. (Error: PortInUseError: Port 18800 is already in use.)"

All functional browser tool calls (snapshot, screenshot, act, navigate) then fail for the remainder of the gateway session. Gateway restart is required to recover.

### OpenClaw version

2026.2.18

### Operating system

Linux — Debian 12 (bookworm), x86_64

### Install method

Nix (home-manager, headless server, no sandbox: noSandbox: true)

### Logs, screenshots, and evidence

```shell
Gateway log pattern (repeatable):

[browser/chrome] 🦞 openclaw browser started (custom) profile "openclaw" on 127.0.0.1:18800 (pid 2692478)
[tools] browser failed: Can't reach the OpenClaw browser control service. ... (Error: PortInUseError: Port 18800 is already in use.)

Port check after start/open:
$ ss -tlnp | grep 18800
LISTEN 0 10 127.0.0.1:18800 0.0.0.0:* users:(("chromium",pid=2692478,fd=64))

Workaround that works:
1. pkill -9 chromium
2. Call browser(action="screenshot", targetUrl="...", profile="openclaw")  ← bootstraps control service's own Chromium
3. Call browser(action="tabs", ...) to get targetId
4. Call browser(action="navigate", targetId=..., targetUrl="...")
5. Call browser(action="snapshot", targetId=...) ← works
```

### Impact and severity

Affected: Any agent on Linux using browser(action="start") or browser(action="open") as an initialization step (which the tool description implies is the correct entry point)
Severity: High — blocks all browser automation until gateway restart
Frequency: 100% reproducible
Consequence: Browser tool is effectively unusable unless the agent happens to know the workaround; the documented/intuitive usage pattern (start → open → snapshot) reliably breaks it

### Additional information

The workaround is: never call start or open; instead bootstrap with browser(action="screenshot", targetUrl="...") which lets the browser control service manage Chromium itself. start and open appear to use a different code path that spawns a competing Chromium on port 18800.

Gateway restart also clears the stuck state, but is obviously disruptive.

The tool description in the system prompt currently recommends "start" and "open" for initialization — this should either be corrected or the underlying conflict fixed.

### PR DESCRIPTION
**Pull request by AaronWander**
## **Summary**

- Problem: browser(action="start")/browser(action="open") can return before Chrome’s CDP HTTP+WS is actually ready, so a follow-up snapshot/screenshot races and tries to launch again on the same port, triggering PortInUseError (#21149).
- Why it matters: After the race, browser tool calls can stay broken for the rest of the gateway session (until restart), making browser automation unreliable.
- What changed: After launching local Chrome, ensureBrowserAvailable() now polls briefly for CDP readiness before returning; on timeout it cleans up and fails with a clear error. Added a regression test.
- What did NOT change (scope boundary): No changes to port selection, SSRF policy, extension driver behavior, or snapshot/screenshot output.

## **Change Type (select all)**

- [x]  Bug fix
- [ ]  Feature
- [ ]  Refactor
- [ ]  Docs
- [ ]  Security hardening
- [ ]  Chore/infra

## **Scope (select all touched areas)**

- [ ]  Gateway / orchestration
- [x]  Skills / tool execution
- [ ]  Auth / tokens
- [ ]  Memory / storage
- [ ]  Integrations
- [ ]  API / contracts
- [ ]  UI / DX
- [ ]  CI/CD / infra

## **Linked Issue/PR**

- Closes #21149
- Related #

## **User-visible / Behavior Changes**

- browser(action="start") and browser(action="open") are less likely to cause PortInUseError races with subsequent snapshot/screenshot calls because start/open now waits for CDP readiness before returning.

## **Security Impact (required)**

- New permissions/capabilities? (No)
- Secrets/tokens handling changed? (No)
- New/changed network calls? (No)
- Command/tool execution surface changed? (No)
- Data access scope changed? (No)

### **Environment**

- OS: macOS (dev)
- Runtime/container: Node 22+
- Model/provider: N/A
- Integration/channel (if any): N/A
- Relevant config (redacted): N/A

### **Steps**

1. Start browser via browser(action="start") (or browser(action="open"))
2. Immediately call browser(action="snapshot") / browser(action="screenshot")
3. Observe that the follow-up no longer races into a second launch attempt on the same port

### **Expected**

- No PortInUseError race; follow-up calls work without requiring a gateway restart.

### **Actual**

- Previously could fail with PortInUseError: Port 18800 is already in use and remain broken until restart.

## **Evidence**

- [x]  Failing test/log before + passing after

## **Human Verification (required)**

- Verified scenarios: ran unit test + full repo pnpm check
- Edge cases checked: CDP not ready immediately after launch (polling path)
- What you did **not** verify: live Linux reproduction with real Chrome process timing / real gateway session

## **Compatibility / Migration**

- Backward compatible? (Yes)
- Config/env changes? (No)
- Migration needed? (No)

## **Failure Recovery (if this breaks)**

- Revert this PR commit.
- Known bad symptoms: browser start/open taking slightly longer, or start failing with “CDP websocket … not reachable after start”.

## **Risks and Mitigations**

- Risk: start/open waits slightly longer on slow machines.
    - Mitigation: bounded polling with a small max wait; only applies to the “we just launched Chrome” path.

### CODE REVIEW

File: src/browser/server-context.ts [hunk 1]
```diff
@@ -281,6 +281,21 @@ function createProfileContext(
     const isExtension = profile.driver === "extension";
     const profileState = getProfileState();
     const httpReachable = await isHttpReachable();
+    const waitForCdpReadyAfterLaunch = async () => {
+      // launchOpenClawChrome() can return before Chrome is fully ready to serve /json/version + CDP WS.
+      // If a follow-up call (snapshot/screenshot/etc.) races ahead, we can hit PortInUseError trying to
+      // launch again on the same port. Poll briefly so browser(action="start"/"open") is stable.
+      const maxAttempts = 50;
+      for (let attempt = 0; attempt < maxAttempts; attempt++) {
+        if (await isReachable(1200)) {
```
[the remark this answers is not in the document]
Comment (AaronWander) [reply src/browser/server-context.ts]: Thanks — good catch. The original post-launch loop was bounded by attempt count, but each attempt could block on isReachable(1200) (loopback WS handshake), so in the worst case it could stall browser(action="start"/"open") far longer than intended.
I updated the wait to be bounded by a wall-clock deadline (~8s) and reduced per-attempt timeouts, so startup failures fail fast while still avoiding the follow-up PortInUseError race. Added/adjusted the regression test to cover the post-launch readiness wait behavior.
